scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+        """
+        Remove the csv after it has been uploaded so that there isn't a deluge of csv files in the local directory.
+        :param filename: The filename of the file to delete.
+        :return:
+        """
+        os.remove(filename)
+
+    # gcs (cloud storage) interactions
+    def put_csv_in_bucket(self, filename: str) -> str:
+        """
+        Puts a local file into a GCS bucket accessible by Jade.
+        :param filename: The filename of the file to upload.
+        :return: The gs-path of the uploaded file.
+        """
+        storage_client = storage.Client(project=self.bucket_project, credentials=self.gcp_creds)
+        bucket = storage_client.bucket(self.bucket)
+        blob = bucket.blob(filename)
+        blob.upload_from_filename(filename)
+
+        filepath = f"gs://{self.bucket}/{filename}"
+        print(f"Put a soft-delete file here: {filepath}")
+
+        return filepath
+
+    # jade interactions
+    @staticmethod
+    @lru_cache()
+    def _get_endpoint(endpoint: str):
+        """
+        Get relevant information, such as the HTTP method, the path, and parameters, for Jade API endpoints.
+        :param endpoint: The endpoint name to get information for.
+        :return: A dictionary containing the endpoint's method, path, and parameters.
+        """
+        url = "https://jade-terra.datarepo-prod.broadinstitute.org/v2/api-docs"
+        response = CachedSession().get(url=url)
+        paths = response.json()["paths"]
+
+        my_map = {}
+
+        for path, methods in paths.items():
+            for method, rest in methods.items():
+                if "parameters" in rest:
+                    params = [(param["name"], param["in"]) for param in rest["parameters"]]
+                else:
+                    params = []
+                my_map.update({rest["operationId"]: {"method": method.upper(), "path": path, "params": params}})
+        if endpoint in my_map:
+            return my_map[endpoint]
+        else:
+            raise KeyError(f"Endpoint named {endpoint} not found!")
+
+    def _hit_jade(self, endpoint, handle_ok, body=None, params=None, query=None):
+        """
+        A generic function for hitting the Jade API.
+        :param endpoint: The endpoint to interact with.
+        :param handle_ok: A function that handles the response from the server if the response is a 2xx code.
+        :param body: An optional body for the API request if needed.
+        :param params: Optional endpoint parameters (not query filters) for the API request if needed.
+        :param query: Optional query parameters for the API request if needed.
+        :return: The result of running handle_ok on the server's response, or an HTTPError.
+        """
+        ep_info = self._get_endpoint(endpoint)
+        url = f"{self.base_url}{ep_info['path']}"
+
+        if params:
+            url = url.format(**params)
+        if query:
+            url = f"{url}/?{urllib.parse.urlencode(query)}"
```

Comment:
i see a lot of potential footguns here if someone, e.g., puts a forward slash at the end of the `url` argument or mixes up `params` and `query`. could we use python's built-in `URL` class to help us generate the URL here?